### PR TITLE
SI-6035: Specialization and separate compilation.

### PR DESCRIPTION
**Pull request by gkossakowski**
Fix an issue when specialization wouldn't work properly
in a setting where separate compilation was involved.
E.g. if you inherit from a class that is specialized and
has been compiled separately you might not get proper
forwarders generated. A test-case that this commit
adds covers that scenario.

The root cause of this issue was the fact that
SPECIALIZED flag was not pickled. Thanks to recent
Flags reorganization (see a9b85db) all that needs
to be done is to set the flag before pickling.

We choose to that in superaccessors phase because
it's a phase that runs before pickling and after
type checking (so we can check if given symbol
has an annotation).

Review by @axel22 and @odersky.

### CODE REVIEW

**Comment by axel22** [comment]
Btw:

https://github.com/gkossakowski/scala/blob/5e1e126a540ac2d5a7e067d396c1ae473b0f2cb6/src/compiler/scala/tools/nsc/transform/UnCurry.scala#L567

I think we can remove the check in `UnCurry` now, no?

File: src/compiler/scala/tools/nsc/typechecker/SuperAccessors.scala [hunk 1]
```diff
@@ -208,6 +212,11 @@ abstract class SuperAccessors extends transform.Transform with transform.TypingT
         case TypeApply(sel @ Select(This(_), name), args) =>
           mayNeedProtectedAccessor(sel, args, false)
 
+        // set a flag for all type parameters with `@specialized` annotation so it can be pickled
+        case typeDef: TypeDef if typeDef.symbol.hasAnnotation(definitions.SpecializedClass) =>
```
Comment (VladUreche) [inline src/compiler/scala/tools/nsc/typechecker/SuperAccessors.scala:216]: Greg, you set the SPECIALIZED flag here and it gets picked, but how do you know what value types the type param is specialized on? You need to read the annotation again, right? If so, why do you need the flag, couldn't you just read the annotation again?

Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/typechecker/SuperAccessors.scala:216]: Vlad, you are right. Both the flag and annotation gets pickled. Once you check the flag is set you read the annotation. The point of doing that is to have very quick check (by using a flag) for short-cutting. Does it make sense?

**Comment by gkossakowski** [comment]
@axel22: You are right. I totally forgot. The commit arrived.

**Comment by axel22** [comment]
lgtm!

**Comment by gkossakowski** [comment]
Is jenkins going to recheck this pull request automatically?

**Comment by gkossakowski** [comment]
FYI: I'm looking into that failure now.

**Comment by gkossakowski** [comment]
Adressed all feedback (including failing tests) in new pull request: https://github.com/scala/scala/pull/915

Tests were failing because I was setting a flag on a type parameter skolem and not type parameter itself.

**Pull request closed without merging**